### PR TITLE
feat(ip-restriction) add support for IPv6

### PR DESCRIPTION
**Pull request by murillopaula**
Add IPv6 support for the ip-restriction plugin.

Looking for feedback so I can see if it would be a good idea to replace the 'lua-resty-iputils' and 'lua-resty-mediador' libraries by the 'lua-resty-ipmatcher' one.

The iputils lib is only used for typedef validations and the mediador lib is used for trusted ips on the pdk side and for sources and destinations on the router side.

- [feat(ip-restriction) add support for IPv6](https://github.com/Kong/kong/pull/5640/commits/fb76a01bfcea55e50353402204be6b2c62eebd3a)

### CODE REVIEW

File: kong/db/schema/typedefs.lua [hunk 1]
```diff
@@ -75,6 +76,17 @@ local function validate_cidr_v4(ip)
 end
 
 
+local function validate_cidr_v4_v6(ip)
```
Comment (bungle) [inline kong/db/schema/typedefs.lua]: How is this different from `local function validate_ip_or_cidr(ip)` above? I know it uses different library, but both accept `ip` and `cidr`. So could you just use the `typedefs.ip_or_cidr` with `ip-restriction` and skip adding this for now? We will rework these when removing other deps.

File: kong/plugins/ip-restriction/handler.lua [hunk 2]
```diff
@@ -1,68 +1,47 @@
...
+  local err = false
   local binary_remote_addr = ngx.var.binary_remote_addr
 
   if not binary_remote_addr then
     return kong.response.exit(FORBIDDEN, { message = "Cannot identify the client IP address, unix domain sockets are not supported." })
   end
 
   if conf.blacklist and #conf.blacklist > 0 then
-    block = iputils.binip_in_cidrs(binary_remote_addr, cidr_cache(conf.blacklist))
+    if not kong.ctx.plugin.blacklist then
```
Comment (bungle) [inline kong/plugins/ip-restriction/handler.lua]: When would this hold `true`?

File: kong/plugins/ip-restriction/handler.lua [hunk 3]
```diff
@@ -1,68 +1,47 @@
...
+      kong.ctx.plugin.blacklist = ipmatcher.new(conf.blacklist)
+    end
+
+    block, err = kong.ctx.plugin.blacklist:match_bin(binary_remote_addr)
   end
 
   if conf.whitelist and #conf.whitelist > 0 then
-    block = not iputils.binip_in_cidrs(binary_remote_addr, cidr_cache(conf.whitelist))
+    if not kong.ctx.plugin.whitelist then
```
Comment (bungle) [inline kong/plugins/ip-restriction/handler.lua]: When would this hold `true`?

File: kong/plugins/ip-restriction/handler.lua [hunk 4]
```diff
@@ -1,68 +1,47 @@
...
   local binary_remote_addr = ngx.var.binary_remote_addr
 
   if not binary_remote_addr then
     return kong.response.exit(FORBIDDEN, { message = "Cannot identify the client IP address, unix domain sockets are not supported." })
   end
 
   if conf.blacklist and #conf.blacklist > 0 then
-    block = iputils.binip_in_cidrs(binary_remote_addr, cidr_cache(conf.blacklist))
+    if not kong.ctx.plugin.blacklist then
+      kong.ctx.plugin.blacklist = ipmatcher.new(conf.blacklist)
```
Comment (bungle) [inline kong/plugins/ip-restriction/handler.lua]: What is the reason to set the context value?

File: kong/plugins/ip-restriction/handler.lua [hunk 5]
```diff
@@ -1,68 +1,47 @@
...
+    end
+
+    block, err = kong.ctx.plugin.blacklist:match_bin(binary_remote_addr)
   end
 
   if conf.whitelist and #conf.whitelist > 0 then
-    block = not iputils.binip_in_cidrs(binary_remote_addr, cidr_cache(conf.whitelist))
+    if not kong.ctx.plugin.whitelist then
+      kong.ctx.plugin.whitelist = ipmatcher.new(conf.whitelist)
```
Comment (bungle) [inline kong/plugins/ip-restriction/handler.lua]: What is the reason to set the context value?

File: kong/plugins/ip-restriction/schema.lua [hunk 6]
```diff
@@ -8,8 +8,8 @@ return {
     { config = {
         type = "record",
         fields = {
-          { whitelist = { type = "array", elements = typedefs.cidr_v4, }, },
-          { blacklist = { type = "array", elements = typedefs.cidr_v4, }, },
+          { whitelist = { type = "array", elements = typedefs.cidr_v4_v6, }, },
```
Comment (bungle) [inline kong/plugins/ip-restriction/schema.lua]: Here as I commented above, you could just use:
```
typedefs.ip_or_cidr
```

File: kong/plugins/ip-restriction/schema.lua [hunk 7]
```diff
@@ -8,8 +8,8 @@ return {
...
+          { blacklist = { type = "array", elements = typedefs.cidr_v4_v6, }, },
```
Comment (bungle) [inline kong/plugins/ip-restriction/schema.lua]: Here as I commented above, you could just use:
```
typedefs.ip_or_cidr
```

File: kong/plugins/ip-restriction/handler.lua (hunk 2 above)
Comment (murillopaula) [reply kong/plugins/ip-restriction/handler.lua]: `kong.ctx.*` should not be used to cache `ipmatcher` object instances

File: kong/plugins/ip-restriction/handler.lua (hunk 3 above)
Comment (murillopaula) [reply kong/plugins/ip-restriction/handler.lua]: `kong.ctx.*` should not be used to cache `ipmatcher` object instances

File: kong/plugins/ip-restriction/handler.lua (hunk 4 above)
Comment (murillopaula) [reply kong/plugins/ip-restriction/handler.lua]: `kong.ctx.*` should not be used to cache `ipmatcher` object instances

File: kong/plugins/ip-restriction/handler.lua (hunk 5 above)
Comment (murillopaula) [reply kong/plugins/ip-restriction/handler.lua]: `kong.ctx.*` should not be used to cache `ipmatcher` object instances

File: kong/db/schema/typedefs.lua (hunk 1 above)
Comment (murillopaula) [reply kong/db/schema/typedefs.lua]: we can break existing third party plugins using `validate_ip_or_cidr`

the new `ipmatcher` lib doesn't validate stuff like this:

```
{ whitelist = { "0.0.0.0/a", "0.0.0.0/-1", "0.0.0.0/33" }
```

i.e. `ipmatcher` doesn't complain on wrong cidrs... this issue has been reported at https://github.com/api7/lua-resty-ipmatcher/issues/11

we can also break the error message, but that is a very minor problem

File: kong/plugins/ip-restriction/schema.lua (hunk 6 above)
Comment (murillopaula) [reply kong/plugins/ip-restriction/schema.lua]: we can break existing third party plugins using `validate_ip_or_cidr` (look at my comment above)

File: kong/plugins/ip-restriction/schema.lua (hunk 7 above)
Comment (murillopaula) [reply kong/plugins/ip-restriction/schema.lua]: we can break existing third party plugins using validate_ip_or_cidr (view comment above)

File: kong/db/schema/typedefs.lua (hunk 1 above)
Comment (bungle) [reply kong/db/schema/typedefs.lua]: I don't think we break any existing plugins that use `validate_ip_or_cidr` as we ARE NOT CHANGING it. That typedefs.lua stays the same. This already catches invalid cidrs. We just change the ip-restriction plugin to use `validate_ip_or_cidr`. That validation function DOES NOT use `ipmatcher` but that mediador, and that is fine. The implementation inside `handler.lua` uses `ipmatcher`.

File: kong/plugins/ip-restriction/schema.lua (hunk 6 above)
Comment (bungle) [reply kong/plugins/ip-restriction/schema.lua]: no we cannot as WE ARE NOT CHANGING typedef. Only the schema of ip-restriction plugin.

File: kong/db/schema/typedefs.lua [hunk 8]
```diff
@@ -75,6 +76,17 @@ local function validate_cidr_v4(ip)
...
+  local _, err = ipmatcher.new({ ip })
+
```
Comment (bungle) [inline kong/db/schema/typedefs.lua]:
```suggestion
```

File: kong/plugins/ip-restriction/handler.lua [hunk 9]
```diff
@@ -1,68 +1,44 @@
...
 
   if not binary_remote_addr then
     return kong.response.exit(FORBIDDEN, { message = "Cannot identify the client IP address, unix domain sockets are not supported." })
   end
 
   if conf.blacklist and #conf.blacklist > 0 then
-    block = iputils.binip_in_cidrs(binary_remote_addr, cidr_cache(conf.blacklist))
+    ip = ipmatcher.new(conf.blacklist)
+
+    block, err = ip:match_bin(binary_remote_addr)
```
Comment (bungle) [inline kong/plugins/ip-restriction/handler.lua]: Pseudo:
`if err then return kong.response.exit(500, ...) end`

File: kong/plugins/ip-restriction/handler.lua [hunk 10]
```diff
@@ -1,68 +1,44 @@
...
   end
 
   if conf.whitelist and #conf.whitelist > 0 then
-    block = not iputils.binip_in_cidrs(binary_remote_addr, cidr_cache(conf.whitelist))
+    ip = ipmatcher.new(conf.whitelist)
+
+    block, err = ip:match_bin(binary_remote_addr)
```
Comment (bungle) [inline kong/plugins/ip-restriction/handler.lua]: Pseudo:
`if err then return kong.response.exit(500, ...) end`

File: kong/plugins/ip-restriction/handler.lua [hunk 11]
```diff
@@ -1,68 +1,44 @@
...
+    block = not block
+  end
+
+  if err then
+    block = true
+    kong.log.err("invalid binary IP address: " .. err)
   end
```
Comment (bungle) [inline kong/plugins/ip-restriction/handler.lua]: remove

File: kong/plugins/ip-restriction/schema.lua (hunk 6 above)
Comment (bungle) [inline kong/plugins/ip-restriction/schema.lua]: Again, just use `typedefs.ip_or_cidr` here.

Comment (bungle) [inline kong/plugins/ip-restriction/schema.lua]: Again, just use `typedefs.ip_or_cidr` here.

File: kong-2.0.2-0.rockspec [hunk 12]
```diff
@@ -37,6 +37,7 @@ dependencies = {
   "lua-messagepack == 0.5.2",
   "lua-resty-openssl == 0.5.3",
   "lua-resty-counter == 0.2.0",
+  "lua-resty-ipmatcher == 0.5",
```
Comment (bungle) [inline kong-2.0.2-0.rockspec]: They just released `0.6`, so could we update to that too :-(. Sorry.

**Comment by murillopaula** [comment]
Thank you @bungle for reviewing! You are awesome. Thanks for identifying problems and proposing solutions so fast.

(First GitHub comment sent from the GitHub Android app)

**Pull request merged**